util/awsservice/cloudwatchlogs.go

```diff
@@ -161,15 +168,97 @@ func GetLogStreams(logGroupName string) []types.LogStream {
 	return []types.LogStream{}
 }
 
-func MatchEMFLogWithSchema(logEntry string, s *jsonschema.Schema, logValidator func(string) bool) bool {
-	keyErrors, e := s.ValidateBytes(context.Background(), []byte(logEntry))
-	if e != nil {
-		log.Println("failed to execute schema validator:", e)
-		return false
-	} else if len(keyErrors) > 0 {
-		log.Printf("failed schema validation: %v\n", keyErrors)
-		return false
+type LogEventValidator func(event types.OutputLogEvent) error
+
+type LogEventsValidator func(events []types.OutputLogEvent) error
+
+type SchemaRetriever func(message string) (string, error)
+
+func WithSchema(schema string) SchemaRetriever {
+	return func(_ string) (string, error) {
+		return schema, nil
+	}
+}
+
+func AssertLogSchema(schemaRetriever SchemaRetriever) LogEventValidator {
+	return func(event types.OutputLogEvent) error {
+		message := *event.Message
+		if schemaRetriever == nil {
+			return errors.New("nil schema retriever")
+		}
+		schema, err := schemaRetriever(*event.Message)
+		if err != nil {
+			return fmt.Errorf("unable to retrieve schema: %w", err)
+		}
+		keyErrors, err := jsonschema.Must(schema).ValidateBytes(context.Background(), []byte(message))
+		if err != nil {
+			return fmt.Errorf("failed to execute schema validator: %w", err)
+		} else if len(keyErrors) > 0 {
+			return fmt.Errorf("failed schema validation: %v | schema: %s | log: %s", keyErrors, schema, message)
+		}
+		return nil
+	}
+}
+
+func AssertLogSubstring(substr string) LogEventValidator {
+	return func(event types.OutputLogEvent) error {
+		if !strings.Contains(*event.Message, substr) {
+			return fmt.Errorf("log event message missing substring (%s): %s", substr, *event.Message)
+		}
+		return nil
 	}
+}
+
+// AssertPerLog runs each validator on each of the log events. Fails fast.
+func AssertPerLog(validators ...LogEventValidator) LogEventsValidator {
+	return func(events []types.OutputLogEvent) error {
+		for _, event := range events {
+			for _, validator := range validators {
+				if err := validator(event); err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	}
+}
+
+func AssertLogsNotEmpty() LogEventsValidator {
+	return func(events []types.OutputLogEvent) error {
+		if len(events) == 0 {
+			return errors.New("no log events")
+		}
+		return nil
+	}
+}
 
-	return logValidator(logEntry)
+func AssertLogsCount(count int) LogEventsValidator {
+	return func(events []types.OutputLogEvent) error {
+		if len(events) != count {
+			return fmt.Errorf("actual log events count (%v) does not match expected (%v)", len(events), count)
+		}
+		return nil
+	}
+}
+
+func AssertNoDuplicateLogs() LogEventsValidator {
+	return func(events []types.OutputLogEvent) error {
+		byTimestamp := make(map[int64]map[string]struct{})
```

Comment:
I guess there could be an agent bug reporting the same event multiple times with different timestamps... Like rereading a file or rereading an event log. But it is hard (or impossible) to distinguish this scenario from an event that legitimately repeats at different times.

Reply:
Yeah we repeat the same logs quite often for tests, so it'd be harder to catch that.